### PR TITLE
Fixes some openssl and libressl compatability issues

### LINKED ISSUE #48: Investigate reports on failing tests from CPAN testers; 1.92 (OpenBSD)
Investigate reports on failing tests from CPAN testers. Seems to be OpenBSD issues.

Crypt-OpenSSL-PKCS12-1.92:

- OpenBSD.amd64-openbsd / 5.38.0:
  - FAIL http://www.cpantesters.org/cpan/report/f41dbe04-72f9-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.38.2:
  - FAIL http://www.cpantesters.org/cpan/report/633b5b36-727b-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.38.2:
  - FAIL http://www.cpantesters.org/cpan/report/f5687540-72ba-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.10.0:
  - FAIL http://www.cpantesters.org/cpan/report/52052060-7189-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.40.0:
  - FAIL http://www.cpantesters.org/cpan/report/5f108be4-71f4-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.8.9:
  - FAIL http://www.cpantesters.org/cpan/report/957a7858-71b2-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.10.0:
  - FAIL http://www.cpantesters.org/cpan/report/d2e54c0a-71a2-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.40.0:
  - FAIL http://www.cpantesters.org/cpan/report/69a98b6a-7239-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.8.9:
  - FAIL http://www.cpantesters.org/cpan/report/cb6aaa0e-71c2-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.10.1:
  - FAIL http://www.cpantesters.org/cpan/report/7193bc3e-7152-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.12.0:
  - FAIL http://www.cpantesters.org/cpan/report/583e9d9c-710e-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.12.1:
  - FAIL http://www.cpantesters.org/cpan/report/b02784fa-70bf-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.10.1:
  - FAIL http://www.cpantesters.org/cpan/report/24a6eb98-716c-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.12.0:
  - FAIL http://www.cpantesters.org/cpan/report/16fd9c74-7132-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.12.1:
  - FAIL http://www.cpantesters.org/cpan/report/45db1a14-70ea-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.12.2:
  - FAIL http://www.cpantesters.org/cpan/report/d6fdecdc-706e-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.12.3:
  - FAIL http://www.cpantesters.org/cpan/report/84c0c282-701e-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.12.2:
  - FAIL http://www.cpantesters.org/cpan/report/170554e2-709b-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.12.3:
  - FAIL http://www.cpantesters.org/cpan/report/97b6c302-7049-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.12.4:
  - FAIL http://www.cpantesters.org/cpan/report/a903b412-6ff9-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.12.4:
  - FAIL http://www.cpantesters.org/cpan/report/fc7d38f8-6fd0-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.12.5:
  - FAIL http://www.cpantesters.org/cpan/report/2da8396e-6f81-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.14.0:
  - FAIL http://www.cpantesters.org/cpan/report/afd27b52-6f22-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.12.5:
  - FAIL http://www.cpantesters.org/cpan/report/63bb2f34-6fa9-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd-thread-multi / 5.14.0:
  - FAIL http://www.cpantesters.org/cpan/report/490940f4-6f59-11ef-98b0-b3c3213a625c 

- OpenBSD.amd64-openbsd / 5.14.1:
  - FAIL http://www.cpantesters.org/cpan/report/ccaac62e-6eb6-11ef-98b0-b3c3213a625c

- OpenBSD.amd64-openbsd-thread-multi / 5.14.1:
  - FAIL http://www.cpantesters.org/cpan/report/c496dcea-6eeb-11ef-98b0-b3c3213a625c

- OpenBSD.amd64-openbsd-thread-multi / 5.14.2:
  - FAIL http://www.cpantesters.org/cpan/report/34f1c456-6e7d-11ef-98b0-b3c3213a625c

### LINKED ISSUE #49: Investigate reports on failing tests from CPAN testers; 1.92 (Windows)
Investigate reports on failing tests from CPAN testers. Windows issue:

Crypt-OpenSSL-PKCS12-1.92:

- MSWin32-x64-multi-thread / 5.40.0:
  - FAIL http://www.cpantesters.org/cpan/report/ffa00e1e-6dd8-1014-9fa9-da9af891d987 

OpenBSD has a lot of reports on failing tests, see: #48

### LINKED ISSUE #50: Investigate reports on failing tests from CPAN testers; 1.92 (Dragonfly)
Crypt-OpenSSL-PKCS12-1.92:
- x86_64-dragonfly / 5.41.4:
  - FAIL http://www.cpantesters.org/cpan/report/ad1649d8-75e0-11ef-b22b-818eedf47c6a 

- x86_64-dragonfly-thread-multi / 5.41.4:
  - FAIL http://www.cpantesters.org/cpan/report/6a99fb0c-7583-11ef-aa02-c67b1057315a

### PR DESCRIPTION
**Pull request by timlegge**
# Description

Fixes some compatibility issues recently logged. 

Fixes/addresses (If applicable) # (issue)
Fixes #48
Fixes #49
Fixes #50

## Type of change

Please delete options that are not relevant.

- [ X] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules
- [X] Changes have been manually tested (please provide information on test platform using the fields below)

## Test / Development Platform Information

- Operating system and version
- Crypt::OpenSSL::PKCS12 version master
- Perl version various
- OpenSSL and LibreSSL on DragonBSD, OpenBSD

### CODE REVIEW

File: Makefile.PL [hunk 1]
```diff
@@ -20,7 +20,7 @@ my $cc_option_flags = " -O2 -g";
 if ($Config::Config{cc} =~ /gcc/i) {
   $cc_option_flags .= $ENV{AUTHOR_TESTING} ? ' -Wall -Werror' : ' -Wall';
 } else {
-  $cc_option_flags .= '';
+  $cc_option_flags .= ' -Wno-compound-token-split-by-macro';
```
Comment (timlegge) [inline Makefile.PL:23]: Need to look at where to add this.  My only be needed on freebsd or openbsd

File: PKCS12.pm [hunk 2]
```diff
@@ -71,6 +71,10 @@ This distribution implements a subset of OpenSSL's PKCS12 API.
 
 =item * new( )
 
+=item * legacy_support ( )
+
+Check whether the openssl version installed can load the legacy provider.
```
Comment (timlegge) [inline PKCS12.pm:76]: reword to has loaded maybe

File: PKCS12.xs [hunk 3]
```diff
@@ -13,6 +13,8 @@
 #include <openssl/ssl.h>
 #if OPENSSL_VERSION_NUMBER >= 0x30000000L
 #include <openssl/provider.h>
+  OSSL_PROVIDER *legacy = NULL;
```
Comment (timlegge) [inline PKCS12.xs:16]: This makes it global - should look at whether it should just be part of the object - as in legacy_loaded

File: PKCS12.xs [hunk 4]
```diff
@@ -898,6 +900,7 @@ static int alg_print(pTHX_ BIO *bio, CONST_X509_ALGOR *alg, HV * parameters_hash
                        ASN1_INTEGER_get(kdf->iter), OBJ_nid2sn(prfnid));
       PBKDF2PARAM_free(kdf);
 #if OPENSSL_VERSION_NUMBER > 0x10100000L
+#ifndef LIBRESSL_VERSION_NUMBER
```
Comment (timlegge) [inline PKCS12.xs:903]: openssl versions differences were bad enough... libressl supports a pre 3.0 version of openssl api

File: PKCS12.xs [hunk 5]
```diff
@@ -992,6 +996,18 @@ new(class)
   OUTPUT:
   RETVAL
 
+IV legacy_support(class)
```
Comment (timlegge) [inline PKCS12.xs:999]: Again, maybe I should just update the object with this - but then again it looks like the Crypt::OpenSSL::PKCS12 is really just the openssl object and does not have any other attributes so this is likely bes

File: PKCS12.xs [hunk 6]
```diff
@@ -992,6 +996,18 @@ new(class)
...
+  SV *class;
+
+  CODE:
+  RETVAL = 0;
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
```
Comment (timlegge) [inline PKCS12.xs:1004]: If OPENSSL_VERSION_NUMBER < 0x30000000 and not libressl it should like default to support for the legacy protocols.

File: t/pkcs12-info-arbitrary-bag-attributes.t
Comment (timlegge) [inline t/pkcs12-info-arbitrary-bag-attributes.t:1]: For the info tests (as a string) I may wrap that test as a AUTHOR only.  It is comparing the result of the modules output to the openssl command on a particular version of openssl and linux.  The hash should not fail as often due to version differences

**Review by timlegge — COMMENTED** [review]
A few comments for me to review @jonasbn

**Comment by timlegge** [comment]
@jonasbn This seems to resolve most things.  Can you review - I am not certain about:

legacy_support - not sure its really necessary it might be better to simply wrap these older PKCS12 file tests in AUTHOR_TESTING

**Review by jonasbn — APPROVED** [review]
Hi @timlegge 

This is approved.

**Comment by jonasbn** [comment]
Hi @timlegge 

In regards to:

> legacy_support - not sure its really necessary it might be better to simply wrap these older > PKCS12 file tests in AUTHOR_TESTING

Let try our your patch and we can also resort to the AUTHOR_TESTING approach

**Pull request merged**